### ISSUE TITLE
Infinite Loop in "Checkin" Function (v5.2)

### ISSUE THREAD

**Issue by 89yoyos** [Aymkdn/SharepointPlus#120]
I've encountered a problem with the Checkin function in the latest version (v5.0) that not only prevents me from checking in a file, but takes the CPU usage of the tab it's running in to 100% until I force-close it. I haven't dug too far into it, but my guess it it's an infinite loop of some sort. It happens regardless of what options I put (even the example version from the documentation triggers it).

The relevant code is:

Some links to SPPlus and a plugin I made for it (to check files out):
```javascript
<script type="text/javascript" src="https://unpkg.com/sharepointplus"></script>
<script type="text/javascript" src="JS/plugin_SPP_CheckOut.js"></script>
```

I define a class `Article` that has some data. Then I use this function to check it in:
```javascript
async checkInArticle(article){
		$SP().checkin({
			destination:"http://www.[redacted].com"+article.ServerURL,
			comments:"Article Upload Automatic Checkin"
		}).then(function() {
			alert("Done");
		}).catch(function(error) {
			alert("Check in failed");
		});
	}
```
(Real URL not included)

I considered that my plugin might be doing something weird, but even if I don't include it in the page, this issue still occurs.

**Comment by Aymkdn** [comment]
Hi.

I've just tested with SharepointPlus v5.2 (the one from `https://unpkg.com/sharepointplus`) and it works as expected.

in your code, because you use `async` you may need to return the result of `$SP().checkin()` : 
```javascript
async checkInArticle(article){
    return $SP().checkin({
	destination:"http://www.[redacted].com"+article.ServerURL,
	comments:"Article Upload Automatic Checkin"
   })
}

[...]
await checkInArticle('/path/to/file.ext');
alert('File Checked In');
```

**Comment by 89yoyos** [comment]
I controlled for that by invoking it directly in the Chrome Developer console command line, too. I've also tried different types of arguments and returns, including the sample you suggested.

**Comment by Aymkdn** [comment]
I'm sure the problem is not from my code because I tested it on a clean page and it worked.

Check that `destination` is correct. Check the dev console to see what's happening and so on...

**Comment by 89yoyos** [comment]
Okay, I just tested it on a blank page and it works there, but the exact same code doesn't work in the main page.

Running this:
```
return $SP().checkin({
	destination:"http://www.[redacted].com/Pages/File.aspx",
	comments:"Article Upload Automatic Checkin"
});
```
It works as expected on a blank page, but doesn't work on the main page. So it is my code, but it's pretty inexplicable. Any suggestions as to why this would happen? I invoke everything in standard ways and don't do anything crazy or modify $SP in any way.

**Comment by Aymkdn** [comment]
You should see a request in the Network tab of your Chrome dev console. What is returned by the server when the Checkin request is sent from SharepointPlus ?

**Comment by 89yoyos** [comment]
As soon as I run the command it hard freezes and the dev console stops working. The average response time for this site is 1-2 seconds for even the most basic requests, so it seems to be freezing as or before it initiates a network request.

**Comment by Aymkdn** [comment]
I don't know... The only action that happens before sending the request is to find the URL.

Maybe you can try : 
```javascript
$SP().checkin({
	url:"http://www.[redacted].com/",
	destination:"http://www.[redacted].com/Pages/File.aspx",
	comments:"Article Upload Automatic Checkin"
});
```

Other than that, I have no idea...

**Comment by 89yoyos** [comment]
Actually, I left it running while I typed that comment and it managed to run out of memory, prompting Chrome to unfreeze long enough for the debugger to stop it and let me step through the program line-by-line.

It gets stuck on line 2422
```javascript
.then(function(url) { setup.url=url; return _this.checkin(setup) })
```
It loops here infinitely.

Adding the URL like you suggest seems to work though.

**Comment by Aymkdn** [comment]
OK. I'll have a closer look for the next release.

**Comment by 89yoyos** [comment]
Thanks for the help

**Comment by Aymkdn** [comment]
You're welcome. I'll keep this issue opened until I'm able to investigate more :)

**Comment by 89yoyos** [comment]
Cool, thanks. 

I'm not sure if this is best put here or in another Issue (trying not to bug you too much lmao) but one thing I noticed now that it's working is that it defaults to checking files in as a major version. This is problematic in sites with Publishing Pages, as checking in a major version also automatically publicly publishes those pages. 

According to the "_vti_bin/Lists.asmx" documentation, if you add the node "CheckinType" to the SOAP request, it would let you pick what type of checkin to do.

For now, I'll probably have to make a plugin to handle this in my script, but that would be a nice addition in the main script.

**Comment by Aymkdn** [comment]
OK, noted for the next version!

I have not the time to work on that before at least a few months, so yeah you may need to do it yourself. You could use https://aymkdn.github.io/SharepointPlus/core.html#webService

**Comment by Aymkdn** [comment]
I've added a `type` option for this `checkin()`, which will be released with SharepointPlus v6.0 in the coming weeks.
